pkg/restore/restore.go

```diff
@@ -605,6 +560,104 @@ func (ctx *restoreContext) execute() (Result, Result) {
 	return warnings, errs
 }
 
+func (ctx *restoreContext) processSelectedResource(
+	selectedResource restoreableResource,
+	totalItems int,
+	i int,
+	existingNamespaces sets.String,
+	update chan progressUpdate,
+) (int, Result, Result) {
+	warnings, errs := Result{}, Result{}
+	groupResource := schema.ParseGroupResource(selectedResource.resource)
+
+	for namespace, selectedItems := range selectedResource.selectedItemsByNamespace {
+		for _, selectedItem := range selectedItems {
+			// If we don't know whether this namespace exists yet, attempt to create
+			// it in order to ensure it exists. Try to get it from the backup tarball
+			// (in order to get any backed-up metadata), but if we don't find it there,
+			// create a blank one.
+			if namespace != "" && !existingNamespaces.Has(selectedItem.targetNamespace) {
+				logger := ctx.log.WithField("namespace", namespace)
+
+				ns := getNamespace(
+					logger,
+					archive.GetItemFilePath(ctx.restoreDir, "namespaces", "", namespace),
+					selectedItem.targetNamespace,
+				)
+				_, nsCreated, err := kube.EnsureNamespaceExistsAndIsReady(
+					ns,
+					ctx.namespaceClient,
+					ctx.resourceTerminatingTimeout,
+				)
+				if err != nil {
+					errs.AddVeleroError(err)
+					continue
+				}
+
+				// Add the newly created namespace to the list of restored items.
+				if nsCreated {
+					itemKey := velero.ResourceIdentifier{
+						GroupResource: kuberesource.Namespaces,
+						Namespace:     ns.Namespace,
+						Name:          ns.Name,
+					}
+					ctx.restoredItems[itemKey] = struct{}{}
+				}
+
+				// Keep track of namespaces that we know exist so we don't
+				// have to try to create them multiple times.
+				existingNamespaces.Insert(selectedItem.targetNamespace)
+			}
+
+			obj, err := archive.Unmarshal(ctx.fileSystem, selectedItem.path)
+			if err != nil {
+				errs.Add(
+					selectedItem.targetNamespace,
+					fmt.Errorf(
+						"error decoding %q: %v",
+						strings.Replace(selectedItem.path, ctx.restoreDir+"/", "", -1),
+						err,
+					),
+				)
+				continue
+			}
+
+			w, e := ctx.restoreItem(obj, groupResource, selectedItem.targetNamespace)
+			warnings.Merge(&w)
+			errs.Merge(&e)
+			i++
```

Comment:
Just to follow up here from one of the comments above, this is the variable that I think shouldn't be passed in, but instead should be initialised in this function, returned, and used to update the other variable in the calling scope.

Reply:
@zubron On the whole, I think all of your suggestions are good ones to improve readability. Much of this is just moving around what was already in place, but the suggestions all make sense. One caveat is that this func actually needs the total processed, since the progress updates are issued from within it. SO we still need the "processed before this call" number (currently i) passed in. However, if it makes it clearer, rather than adding to it and returning it, we could just count within the func and return that, but then when reporting progress, instead of `i` we'll just use `previousRestoredItems+currentRestoredItems`.

Reply:
Ahh, you're right. Sorry, I misunderstood this when reading it yesterday. I think in that case we can still pass it in to avoid changing the existing logic too much, but a more descriptive name would definitely help :)

Reply:
Thinking about it again, I'm realizing that when it was just `i` before my change, while a more descriptive name would have been better, it was less important since it was just a local variable. Now that we're passing it into a func, it's actually pretty important to have a descriptive name. I'll update it.

Reply:
Renaming `i` to `processedItems` here.